README.md
# AMQP Client

## About

This module aims to ease the communication with a broker using the AMQP protocol, it provides some low-level tools.

## Getting started

## Prerequisites

This module requires Node 8.9 (the latest LTS) and a connection to an AMQP broker

## Installation

```javascript
import Client from '@prismamedia/amqp-client';
// or : Client = require('@prismamedia/amqp-client');
```

## Configuration

```javascript
const client = new Client('amqp://rabbitmq/');
```

## Usage

```javascript
const payload = { ... };

try {
  await client.publish('', 'queue_name', payload);
} catch (err) {
  // Handle error
}
```

```javascript
try {
  const consumerId = await client.consume('queue_name', (message, payload, ack) => {
    // Do what you want with the full AMQP message or with the Object payload

    ack();
  });

  // [...]

  await client.stopConsumer(consumerId, true);
} catch (err) {
  // Handle error
}
```


CHANGELOG.md
# AMQP Client

## Changelog

# [6.0.0] (2021-06-29)

### Removed

- Dropped support for Node 12

### Added

- Added support for Node 16

# [4.0.0](https://gitlab.com/prismamediadigital/one/js-amqp-client/compare/3.0.6...4.0.0) (2019-03-27)

### Breaking changes

- Supports only node >= 8.9.0

# [3.0.0](https://bitbucket.org/prismamediadigital/js-amqp-client/branches/compare/3.0.0%0D2.0.3) (2017-09-27)

### Breaking changes

- Removed timeouts in order to simplify the code

# [2.0.3](https://bitbucket.org/prismamediadigital/js-amqp-client/branches/compare/2.0.3%0D2.0.2) (2017-09-10)

### Bugfix

- Fixed flow definitions

# [2.0.2](https://bitbucket.org/prismamediadigital/js-amqp-client/branches/compare/2.0.2%0D2.0.1) (2017-09-08)

### Bugfix

- Fixed resolve/reject order
- Forced onExecutionError callback to be provided

# [2.0.1](https://bitbucket.org/prismamediadigital/js-amqp-client/branches/compare/2.0.1%0D2.0.0) (2017-09-07)

### Bugfix

- Forgot to build the project

# [2.0.0](https://bitbucket.org/prismamediadigital/js-amqp-client/branches/compare/2.0.0%0D1.1.1) (2017-09-07)

### Features

- Allowed timeouts for all the consumers
- Allowed to define the error handlers

# [1.1.1](https://bitbucket.org/prismamediadigital/js-amqp-client/branches/compare/1.1.1%0D1.1.0) (2017-08-29)

### Features

- Semaphores
  - the timeouts are disabled by default.

# [1.1.0](https://bitbucket.org/prismamediadigital/js-amqp-client/branches/compare/1.1.0%0D1.0.4) (2017-08-29)

### Features

- Semaphores
  - allow to configure the following timeouts :
    - lock_timeout : time spent waiting for the "semaphore"
    - execution_timeout : time spent waiting for the end of the provided task, after the "semaphore" is acquired


src/__tests__/client.test.ts
import { Options } from 'amqplib';
import { URL } from 'url';
import { Client } from '../client';
import { ConsumerEventKind, ConsumerStatus } from '../consumer';

type TConsumerRequest = {
  request?: 'waitFor100ms' | 'throw';
};

type TConsumerResponse = {
  response: string;
};

type TRPCConsumerRequest = {
  request?: 'waitFor100ms' | 'throw';
};

type TRPCConsumerResponse = {
  request: TRPCConsumerRequest;
  response: string;
};

async function waitFor(ms: number): Promise<void> {
  await new Promise((resolve) => setTimeout(resolve, ms));
}

describe('Client', () => {
  let client: Client;

  beforeEach(() => {
    client = new Client(process.env.AMQP_DSN!).on('error', console.error);
  });

  afterEach(async () => client.close());

  it('adds some options to the provided URL', () => {
    expect(new URL(client.url).search).toEqual('?heartbeat=15');
  });

  it('forceExchange works', async () => {
    const exchangeName = 'test_force-exchange_works';

    await expect(
      client.assertExchange(exchangeName, 'fanout'),
    ).resolves.toMatchObject({
      exchange: exchangeName,
    });

    await expect(
      client.assertExchange(exchangeName, 'topic'),
    ).rejects.toThrowError();

    await expect(
      client.forceExchange(exchangeName, 'topic'),
    ).resolves.toMatchObject({
      exchange: exchangeName,
    });

    await expect(client.deleteExchange(exchangeName)).resolves.toEqual({});
  });

  it('forceQueue works', async () => {
    const queueName = 'test_force-queue_works';

    await expect(client.assertQueue(queueName)).resolves.toMatchObject({
      queue: queueName,
    });

    await expect(
      client.assertQueue(queueName, { exclusive: true }),
    ).rejects.toThrowError();

    await expect(
      client.forceQueue(queueName, { exclusive: true }),
    ).resolves.toMatchObject({
      queue: queueName,
    });

    await expect(client.deleteQueue(queueName)).resolves.toEqual({
      messageCount: 0,
    });
  });

  it('publishing works', async () => {
    const queueName = 'test_publishing_works';
    const queueOptions: Options.AssertQueue = { exclusive: true };

    await client.forceQueue(queueName, queueOptions);
    await client.purgeQueue(queueName);

    await Promise.all(
      [...new Array(10)].map(() =>
        expect(client.publish('', queueName, {})).resolves.toBeUndefined(),
      ),
    );

    await expect(client.checkQueue(queueName)).resolves.toMatchObject({
      messageCount: 10,
    });

    await expect(client.deleteQueue(queueName)).resolves.toMatchObject({
      messageCount: 10,
    });
  });

  it('consuming works', async () => {
    const queueName = 'test_consuming_works';
    const queueOptions: Options.AssertQueue = { exclusive: true };

    await client.forceQueue(queueName, queueOptions);
    await client.purgeQueue(queueName);

    try {
      let messageProcessedCount = 0;

      const consumer = await client.consume<
        TConsumerRequest,
        TConsumerResponse
      >(
        queueName,
        async ({ payload }) => {
          messageProcessedCount++;

          switch (payload.request) {
            case 'waitFor100ms':
              await waitFor(100);
              break;

            case 'throw':
              throw new Error(`An error`);
          }

          return {
            response: 'OK',
          };
        },
        { prefetch: 1, idleInMs: 250 },
      );

      await Promise.all([
        Promise.all(
          [...new Array(5)].map(() =>
            client.publish<TConsumerRequest>('', queueName, {
              request: 'waitFor100ms',
            }),
          ),
        ),
        Promise.all(
          [...new Array(5)].map(() =>
            client.publish<TConsumerRequest>('', queueName, {
              request: 'throw',
            }),
          ),
        ),
        consumer.wait(ConsumerEventKind.Stopped),
      ]);

      expect(messageProcessedCount).toBe(10);

      await Promise.all(
        [...new Array(5)].map(() =>
          client.publish<TConsumerRequest>('', queueName, {}),
        ),
      );

      await expect(client.checkQueue(queueName)).resolves.toMatchObject({
        messageCount: 5,
      });

      await consumer.startAndWait(ConsumerEventKind.Stopped);

      expect(messageProcessedCount).toBe(15);
    } finally {
      await client.deleteQueue(queueName);
    }
  });

  it('stops on error works', async () => {
    const queueName = 'test_stop_on_error_works';
    const queueOptions: Options.AssertQueue = { exclusive: true };

    await client.forceQueue(queueName, queueOptions);
    await client.purgeQueue(queueName);

    try {
      const consumer = await client.consume(
        queueName,
        () => {
          throw new Error('An error');
        },
        { stopOnError: true },
      );

      await Promise.all([
        [...new Array(5)].map(() => client.publish('', queueName, {})),
        consumer.wait(ConsumerEventKind.Stopped),
      ]);

      // The first message throws an Error, so the consumer is stopped and 4 messages remains
      await expect(client.checkQueue(queueName)).resolves.toMatchObject({
        messageCount: 4,
      });
    } finally {
      await client.deleteQueue(queueName);
    }
  });

  it('stops on signals works', async () => {
    const queueName = 'test_consuming_works';
    const queueOptions: Options.AssertQueue = { exclusive: true };

    await client.forceQueue(queueName, queueOptions);
    await client.purgeQueue(queueName);

    try {
      const consumer = await client.consume<any, void>(
        queueName,
        async () => {},
        { idleInMs: 60000, stopOnSignals: true },
      );

      async function waitThenSendSignal(): Promise<void> {
        await waitFor(100);

        process.emit('SIGINT', 'SIGINT');
      }

      await expect(
        Promise.all([
          consumer.wait(ConsumerEventKind.Stopped),
          waitThenSendSignal(),
        ]),
      ).resolves.toBeTruthy();
    } finally {
      await client.deleteQueue(queueName);
    }
  });

  it('rpc works', async () => {
    const queueName = 'test_rpc_works';
    const queueOptions: Options.AssertQueue = { exclusive: true };

    await client.forceQueue(queueName, queueOptions);
    await client.purgeQueue(queueName);

    const consumer = await client.consume<
      TRPCConsumerRequest,
      TRPCConsumerResponse
    >(queueName, async ({ payload }) => {
      switch (payload.request) {
        case 'waitFor100ms':
          await waitFor(100);
          break;

        case 'throw':
          throw new Error(`An error`);
      }

      return {
        request: payload,
        response: 'OK',
      };
    });

    await expect(
      client.rpc<TRPCConsumerRequest, TRPCConsumerResponse>('', queueName, {}),
    ).resolves.toEqual({
      request: {},
      response: 'OK',
    });

    await expect(
      client.rpc<TRPCConsumerRequest, TRPCConsumerResponse>(
        '',
        queueName,
        { request: 'waitFor100ms' },
        { timeoutInMs: 250 },
      ),
    ).resolves.toEqual({
      request: { request: 'waitFor100ms' },
      response: 'OK',
    });

    // Will throw an error, as the consumer will eventually reject the message, the client will never receive anything
    await expect(
      client.rpc<TRPCConsumerRequest, TRPCConsumerResponse>(
        '',
        queueName,
        { request: 'throw' },
        { timeoutInMs: 50 },
      ),
    ).rejects.toThrow(/The RPC "[^"]+" has reached the timeout of 50ms/);

    expect(consumer.getStatus()).toEqual(ConsumerStatus.Consuming);

    await Promise.all([
      client.deleteQueue(queueName),
      consumer.wait(ConsumerEventKind.Stopped),
    ]);

    expect(consumer.getStatus()).toEqual(ConsumerStatus.Idle);
  });
});


src/consumer.ts
import {
  TEventArgs,
  TEventName,
  TTypedEventEmitterOptions,
  TypedEventEmitter,
} from '@prismamedia/ts-typed-event-emitter';
import type { ConfirmChannel, Message, Options } from 'amqplib';
import { errorMonitor } from 'events';
import { SignalConstants } from 'os';
import { clearTimeout, setTimeout } from 'timers';
import type { Client, TQueueName } from './client';

function computeSpentTimeInMs(since: bigint): number {
  return Math.round(Number(process.hrtime.bigint() - since) / 1000000);
}

export type TSignal = keyof SignalConstants;

export type TConsumerTag = string;

export type TConsumerCallbackArgs<TPayload> = {
  /**
   * The payload
   */
  payload: TPayload;

  /**
   * The whole message
   */
  message: Message;
};

export type TConsumerCallback<TPayload, TResult> = (
  args: TConsumerCallbackArgs<TPayload>,
) => Promise<TResult>;

export enum ConsumerStatus {
  Idle = 'IDLE',
  Consuming = 'CONSUMING',
}

export enum ConsumerEventKind {
  Stopped = 'STOPPED',
  Started = 'STARTED',
  ChannelError = 'CHANNEL_ERROR',
  ChannelClosed = 'CHANNEL_CLOSED',
  ChannelOpened = 'CHANNEL_OPENED',
  MessageParsed = 'MESSAGE_PARSED',
  MessageRejected = 'MESSAGE_REJECTED',
  MessageRequeued = 'MESSAGE_REQUEUED',
  MessageAcknowledged = 'MESSAGE_ACKNOWLEDGED',
  MessageCallbackError = 'MESSAGE_CALLBACK_ERROR',
}

export type TConsumerReply = {
  replyTo: string;
  correlationId: string;
};

export type TConsumerEvents<TPayload, TResult> = {
  [ConsumerEventKind.Stopped]: [];
  [ConsumerEventKind.Started]: [];
  [ConsumerEventKind.ChannelError]: [Error];
  [ConsumerEventKind.ChannelClosed]: [];
  [ConsumerEventKind.ChannelOpened]: [ConfirmChannel];
  [ConsumerEventKind.MessageParsed]: [
    {
      message: Message;
      payload: TPayload;
      tookInMs: number;
    },
  ];
  [ConsumerEventKind.MessageRejected]: [
    {
      message: Message;
      payload?: TPayload;
      error: Error;
      tookInMs: number;
    },
  ];
  [ConsumerEventKind.MessageRequeued]: [
    {
      message: Message;
      payload: TPayload;
      error: Error;
      tookInMs: number;
    },
  ];
  [ConsumerEventKind.MessageAcknowledged]: [
    {
      message: Message;
      payload: TPayload;
      result: TResult;
      reply?: TConsumerReply;
      tookInMs: number;
    },
  ];
  [ConsumerEventKind.MessageCallbackError]: [
    {
      message: Message;
      payload: TPayload;
      error: Error;
      tookInMs: number;
    },
  ];
};

export type TConsumerOptions<TPayload, TResult> = Omit<
  Options.Consume,
  'noLocal'
> & {
  /**
   * @see https://www.rabbitmq.com/consumer-prefetch.html
   */
  prefetch?: number;

  /**
   * After "consumeInMs" ms, the consumer will stop
   */
  consumeInMs?: number;

  /**
   * After "idleInMs" ms with no message, the consumer will stop
   */
  idleInMs?: number;

  /**
   * In case of uncaught "callback error", either requeue (true) or reject (false) the message
   * Default: false
   */
  requeueOnError?: boolean;

  /**
   * In case of uncaught "callback error", either stop (true) or continues (false) the consumer
   * Default: false
   */
  stopOnError?: boolean;

  /**
   * We can try to gracefully stop the consumer on "known" signals
   * - true = ['SIGINT', 'SIGTERM', 'SIGQUIT']
   * - false = []
   *
   * Default: false / none
   */
  stopOnSignals?: boolean | TSignal[];

  /**
   * Add some event listeners
   */
  on?: TTypedEventEmitterOptions<TConsumerEvents<TPayload, TResult>>;
};

export class Consumer<TPayload = any, TResult = any> extends TypedEventEmitter<
  TConsumerEvents<TPayload, TResult>
> {
  #status: ConsumerStatus = ConsumerStatus.Idle;
  #tag: TConsumerTag | null = null;
  #channel: Promise<ConfirmChannel> | null = null;
  #consumeTimeoutId: ReturnType<typeof setTimeout> | null = null;
  #idleTimeoutId: ReturnType<typeof setTimeout> | null = null;
  #signalHandler = (async () => this.stop()).bind(this);

  public readonly prefetch: number;
  public readonly consumeInMs: number | null;
  public readonly idleInMs: number | null;
  public readonly requeueOnError: boolean;
  public readonly stopOnError: boolean;
  public readonly stopOnSignalSet: ReadonlySet<TSignal>;
  public readonly options: Options.Consume;

  public constructor(
    public readonly client: Client,
    public readonly queueName: TQueueName,
    public readonly callback: TConsumerCallback<TPayload, TResult>,
    {
      prefetch = 1,
      consumeInMs,
      idleInMs,
      requeueOnError = false,
      stopOnError = false,
      stopOnSignals = false,
      on,
      ...options
    }: TConsumerOptions<TPayload, TResult> = {},
  ) {
    super(on);

    this.prefetch = prefetch;
    this.consumeInMs = consumeInMs ?? null;
    this.idleInMs = idleInMs ?? null;
    this.requeueOnError = requeueOnError;
    this.stopOnError = stopOnError;
    this.stopOnSignalSet = new Set(
      typeof stopOnSignals === 'boolean'
        ? stopOnSignals
          ? ['SIGINT', 'SIGTERM', 'SIGQUIT']
          : []
        : stopOnSignals,
    );
    this.options = options;

    this.on(ConsumerEventKind.Stopped, () => {
      this.#status = ConsumerStatus.Idle;
      this.#tag = null;
      this.clearConsumeTimeout();
      this.clearIdleTimeout();
      this.clearSignalHandler();
    })
      .on(errorMonitor, () => {
        if (this.isConsuming()) {
          this.emit(ConsumerEventKind.Stopped);
        }
      })
      .on(ConsumerEventKind.ChannelClosed, () => {
        if (this.isConsuming()) {
          this.emit(ConsumerEventKind.Stopped);
        }
      });
  }

  public getStatus(): ConsumerStatus {
    return this.#status;
  }

  public isConsuming(): boolean {
    return this.#status === ConsumerStatus.Consuming;
  }

  public getTag(): TConsumerTag | null {
    return this.#tag;
  }

  public async getChannel(): Promise<ConfirmChannel> {
    if (!this.#channel) {
      const channel = (
        await (this.#channel = new Promise(async (resolve, reject) => {
          try {
            const connection = await this.client.getConnection();
            const channel = await connection.createConfirmChannel();

            resolve(channel);
          } catch (error) {
            reject(error);
          }
        }))
      )
        // The "close" event is always fired after the "error" one
        .on('error', (error: Error) =>
          this.emit(ConsumerEventKind.ChannelError, error),
        )
        .on('close', () => {
          this.#channel = null;

          this.emit(ConsumerEventKind.ChannelClosed);
        });

      await channel.prefetch(this.prefetch);

      this.emit(ConsumerEventKind.ChannelOpened, channel);
    }

    return this.#channel;
  }

  protected ack(
    channel: ConfirmChannel,
    ...args: TEventArgs<
      TConsumerEvents<TPayload, TResult>,
      ConsumerEventKind.MessageAcknowledged
    >
  ): void {
    const [{ message }] = args;

    try {
      channel.ack(message);
      this.emit(ConsumerEventKind.MessageAcknowledged, ...args);
    } catch (error) {
      if (error instanceof Error && error.name === 'IllegalOperationError') {
        this.isConsuming() && this.emit(ConsumerEventKind.Stopped);
      } else {
        throw error;
      }
    }
  }

  protected requeue(
    channel: ConfirmChannel,
    ...args: TEventArgs<
      TConsumerEvents<TPayload, TResult>,
      ConsumerEventKind.MessageRequeued
    >
  ): void {
    const [{ message }] = args;

    try {
      channel.nack(message, false, true);
      this.emit(ConsumerEventKind.MessageRequeued, ...args);
    } catch (error) {
      if (error instanceof Error && error.name === 'IllegalOperationError') {
        this.isConsuming() && this.emit(ConsumerEventKind.Stopped);
      } else {
        throw error;
      }
    }
  }

  protected reject(
    channel: ConfirmChannel,
    ...args: TEventArgs<
      TConsumerEvents<TPayload, TResult>,
      ConsumerEventKind.MessageRejected
    >
  ): void {
    const [{ message }] = args;

    try {
      channel.nack(message, false, false);
      this.emit(ConsumerEventKind.MessageRejected, ...args);
    } catch (error) {
      if (error instanceof Error && error.name === 'IllegalOperationError') {
        this.isConsuming() && this.emit(ConsumerEventKind.Stopped);
      } else {
        throw error;
      }
    }
  }

  protected clearConsumeTimeout(): void {
    if (this.#consumeTimeoutId) {
      clearTimeout(this.#consumeTimeoutId);
      this.#consumeTimeoutId = null;
    }
  }

  protected setConsumeTimeout(): void {
    this.clearConsumeTimeout();

    this.#consumeTimeoutId =
      this.consumeInMs !== null
        ? setTimeout(
            () =>
              this.stop().catch(() => {
                // Do nothing, it's already stopped
              }),
            this.consumeInMs,
          )
        : null;
  }

  protected clearIdleTimeout(): void {
    if (this.#idleTimeoutId) {
      clearTimeout(this.#idleTimeoutId);
      this.#idleTimeoutId = null;
    }
  }

  protected setIdleTimeout(): void {
    this.clearIdleTimeout();

    this.#idleTimeoutId =
      this.idleInMs !== null
        ? setTimeout(
            () =>
              this.stop().catch(() => {
                // Do nothing, it's already stopped
              }),
            this.idleInMs,
          )
        : null;
  }

  protected clearSignalHandler(): void {
    this.stopOnSignalSet.forEach((signal) =>
      process.removeListener(signal, this.#signalHandler),
    );
  }

  protected setSignalHandler(): void {
    this.clearSignalHandler();

    this.stopOnSignalSet.forEach((signal) =>
      process.on(signal, this.#signalHandler),
    );
  }

  public async start(): Promise<void> {
    if (this.isConsuming()) {
      throw new Error(
        `AMQP consumer workflow error: cannot start a "${
          this.#status
        }" consumer`,
      );
    }

    const channel = await this.getChannel();

    const { consumerTag: tag } = await channel.consume(
      this.queueName,
      async (message) => {
        // If the consumer has been cancelled by RabbitMQ, the message callback will be invoked with null.
        if (message === null) {
          this.emit(ConsumerEventKind.Stopped);
        } else {
          const start = process.hrtime.bigint();

          this.setIdleTimeout();

          try {
            if (message.properties.contentType !== 'application/json') {
              // The message's "Content-Type" is not supported, reject it
              return this.reject(channel, {
                message,
                error: new Error(
                  `The Content-Type "${String(
                    message.properties.contentType,
                  )}" is not supported, the message has been rejected`,
                ),
                tookInMs: computeSpentTimeInMs(start),
              });
            }

            let payload: TPayload;

            try {
              payload = JSON.parse(message.content.toString());
            } catch (error) {
              // The message is not valid JSON, reject it
              return this.reject(channel, {
                message,
                error,
                tookInMs: computeSpentTimeInMs(start),
              });
            }

            this.emit(ConsumerEventKind.MessageParsed, {
              message,
              payload,
              tookInMs: computeSpentTimeInMs(start),
            });

            try {
              const result = await this.callback({ message, payload });

              let reply: TConsumerReply | undefined;
              if (
                message.properties.replyTo &&
                message.properties.correlationId
              ) {
                await this.client.publish(
                  '',
                  message.properties.replyTo,
                  result,
                  { correlationId: message.properties.correlationId },
                );

                reply = {
                  replyTo: message.properties.replyTo,
                  correlationId: message.properties.correlationId,
                };
              }

              return this.ack(channel, {
                message,
                payload,
                result,
                reply,
                tookInMs: computeSpentTimeInMs(start),
              });
            } catch (error) {
              this.emit(ConsumerEventKind.MessageCallbackError, {
                message,
                payload,
                error,
                tookInMs: computeSpentTimeInMs(start),
              });

              if (this.stopOnError) {
                await this.stop();
              }

              // The callback has failed, requeue or reject the message
              return this.requeueOnError
                ? this.requeue(channel, {
                    message,
                    payload,
                    error,
                    tookInMs: computeSpentTimeInMs(start),
                  })
                : this.reject(channel, {
                    message,
                    payload,
                    error,
                    tookInMs: computeSpentTimeInMs(start),
                  });
            }
          } catch (error) {
            // Despite our best efforts, an error has not been caught "properly"
            this.emit('error', error);
          }
        }
      },
      this.options,
    );

    this.#status = ConsumerStatus.Consuming;
    this.#tag = tag;
    this.setConsumeTimeout();
    this.setIdleTimeout();
    this.setSignalHandler();

    this.emit(ConsumerEventKind.Started);
  }

  public async stop(): Promise<void> {
    if (!this.isConsuming()) {
      throw new Error(
        `AMQP consumer workflow error: cannot stop a "${
          this.#status
        }" consumer`,
      );
    }

    if (!this.#tag) {
      throw new Error(
        `AMQP consumer workflow error: cannot stop a "${
          this.#status
        }" consumer without a "tag"`,
      );
    }

    const channel = await this.getChannel();
    await channel.cancel(this.#tag);

    this.emit(ConsumerEventKind.Stopped);
  }

  public async startAndWait<
    TName extends TEventName<TConsumerEvents<TPayload, TResult>>,
  >(...names: TName[]) {
    await this.start();

    return this.wait(...names);
  }
}


src/client.ts
import {
  TEventName,
  TTypedEventEmitterOptions,
  TypedEventEmitter,
} from '@prismamedia/ts-typed-event-emitter';
import { ConfirmChannel, connect, Connection, Options } from 'amqplib';
import crypto from 'crypto';
import { errorMonitor } from 'events';
import { clearTimeout, setTimeout } from 'timers';
import { URL } from 'url';
import {
  Consumer,
  ConsumerEventKind,
  TConsumerCallback,
  TConsumerEvents,
  TConsumerOptions,
} from './consumer';

export type TQueueName = string;

export type TCorrelationId = string;

export enum ClientEventKind {
  ConnectionError = 'CONNECTION_ERROR',
  ConnectionClosed = 'CONNECTION_CLOSED',
  ConnectionOpened = 'CONNECTION_OPENED',
  ChannelError = 'CHANNEL_ERROR',
  ChannelClosed = 'CHANNEL_CLOSED',
  ChannelOpened = 'CHANNEL_OPENED',
}

export type TClientEvents = {
  [ClientEventKind.ConnectionError]: [Error];
  [ClientEventKind.ConnectionClosed]: [];
  [ClientEventKind.ConnectionOpened]: [Connection];
  [ClientEventKind.ChannelError]: [Error];
  [ClientEventKind.ChannelClosed]: [];
  [ClientEventKind.ChannelOpened]: [ConfirmChannel];
};

export type TPublishOptions = Omit<
  Options.Publish,
  'contentEncoding' | 'contentType'
>;

export type TRPCOptions = Omit<TPublishOptions, 'expiration'> & {
  timeoutInMs?: number;
};

export type TClientOptions = {
  rpcResultsQueueName?: TQueueName;

  /**
   * Add some event listeners
   */
  on?: TTypedEventEmitterOptions<TClientEvents>;
};

export class Client extends TypedEventEmitter<TClientEvents> {
  #connection: ReturnType<typeof connect> | null = null;
  #channel: Promise<ConfirmChannel> | null = null;
  #rpcMap = new Map<
    TCorrelationId,
    { resolve: (result: any) => void; reject: (error: Error) => void }
  >();
  #rpcConsumer: Consumer;

  public readonly url: string;

  public constructor(
    /**
     * cf: https://www.rabbitmq.com/uri-spec.html
     */
    url: string,
    public readonly options?: TClientOptions,
  ) {
    super(options?.on);

    const providedUrl = new URL(url);

    if (!providedUrl.searchParams.has('heartbeat')) {
      providedUrl.searchParams.set('heartbeat', '15');
    }

    this.url = providedUrl.toString();

    this.#rpcConsumer = new Consumer(
      this,
      options?.rpcResultsQueueName ||
        `rpc_results.${crypto.randomBytes(4).toString('hex')}`,
      async ({ payload, message }) =>
        this.#rpcMap.get(message.properties.correlationId)?.resolve(payload),
      {
        on: {
          [errorMonitor]: (error) =>
            this.#rpcMap.forEach(({ reject }) => reject(error)),
          [ConsumerEventKind.Stopped]: () =>
            this.#rpcMap.forEach(({ reject }) =>
              reject(new Error(`The RPC consumer stopped`)),
            ),
        },
      },
    );
  }

  public async getConnection(): Promise<Connection> {
    if (!this.#connection) {
      const connection = (await (this.#connection = connect(this.url)))
        // The "close" event is always fired after the "error" one
        .on('error', (error: Error) =>
          this.emit(ClientEventKind.ConnectionError, error),
        )
        .on('close', () => {
          this.#connection = null;

          this.emit(ClientEventKind.ConnectionClosed);
        });

      this.emit(ClientEventKind.ConnectionOpened, connection);
    }

    return this.#connection;
  }

  public async getChannel(): Promise<ConfirmChannel> {
    if (!this.#channel) {
      const channel = (
        await (this.#channel = new Promise(async (resolve, reject) => {
          try {
            const connection = await this.getConnection();
            const channel = await connection.createConfirmChannel();

            resolve(channel);
          } catch (error) {
            reject(error);
          }
        }))
      )
        // The "close" event is always fired after the "error" one
        .on('error', (error: Error) =>
          this.emit(ClientEventKind.ChannelError, error),
        )
        .on('close', () => {
          this.#channel = null;

          this.emit(ClientEventKind.ChannelClosed);
        });

      this.emit(ClientEventKind.ChannelOpened, channel);
    }

    return this.#channel;
  }

  public async close(): Promise<void> {
    await (await this.#connection)?.close();
  }

  public async publish<TPayload = any>(
    exchange: string = '',
    routingKey: string = '',
    payload: TPayload,
    options?: TPublishOptions,
  ): Promise<void> {
    // Forbid usage of undefined and null values as they are not "safe" to be stringified and parsed
    if (payload == null) {
      throw new TypeError('The payload has to be a non-null value.');
    }

    const channel = await this.getChannel();

    await new Promise((resolve, reject) =>
      channel.publish(
        exchange,
        routingKey,
        Buffer.from(JSON.stringify(payload)),
        { ...options, contentType: 'application/json' },
        (error) => (error ? reject(error) : resolve(undefined)),
      ),
    );
  }

  public async consume<TPayload = any, TResult = any>(
    queueName: TQueueName,
    callback: TConsumerCallback<TPayload, TResult>,
    options?: TConsumerOptions<TPayload, TResult>,
  ): Promise<Consumer<TPayload, TResult>> {
    const consumer = new Consumer(this, queueName, callback, options);
    await consumer.start();

    return consumer;
  }

  public async consumeAndWait<TPayload = any, TResult = any>(
    queueName: TQueueName,
    callback: TConsumerCallback<TPayload, TResult>,
    options: TConsumerOptions<TPayload, TResult>,
    ...eventNames: TEventName<TConsumerEvents<TPayload, TResult>>[]
  ) {
    const consumer = await this.consume(queueName, callback, options);

    return consumer.wait(...eventNames);
  }

  public async rpc<TRequest = any, TResponse = any>(
    exchange: string = '',
    routingKey: string = '',
    request: TRequest,
    { timeoutInMs = 60000, ...options }: TRPCOptions = {},
  ): Promise<TResponse> {
    if (!this.#rpcConsumer.isConsuming()) {
      await this.assertQueue(this.#rpcConsumer.queueName, { exclusive: true });
      await this.#rpcConsumer.start();
    }

    const correlationId = crypto.randomBytes(4).toString('hex');

    return new Promise(async (resolve, reject) => {
      // Handle the timeout
      const timeoutId = setTimeout(() => {
        this.#rpcMap.delete(correlationId);

        reject(
          new Error(
            `The RPC "${correlationId}" has reached the timeout of ${timeoutInMs}ms`,
          ),
        );
      }, timeoutInMs);

      // Handle the result
      this.#rpcMap.set(correlationId, {
        resolve: (result: TResponse) => {
          clearTimeout(timeoutId);
          this.#rpcMap.delete(correlationId);

          resolve(result);
        },
        reject,
      });

      try {
        await this.publish(exchange, routingKey, request, {
          ...options,
          replyTo: this.#rpcConsumer.queueName,
          correlationId,
          expiration: String(timeoutInMs),
        });
      } catch (error) {
        reject(error);
      }
    });
  }

  /**
   * TODO: Dynamically create these methods
   */
  protected async seamlessCallOnChannel<
    TMethodName extends Extract<
      | 'assertExchange'
      | 'checkExchange'
      | 'bindExchange'
      | 'unbindExchange'
      | 'deleteExchange'
      | 'assertQueue'
      | 'checkQueue'
      | 'bindQueue'
      | 'unbindQueue'
      | 'purgeQueue'
      | 'deleteQueue',
      keyof ConfirmChannel
    >,
  >(
    methodName: TMethodName,
    ...args: Parameters<ConfirmChannel[TMethodName]>
  ): Promise<
    ReturnType<ConfirmChannel[TMethodName]> extends Promise<infer R> ? R : never
  > {
    const channel = await this.getChannel();

    return (channel as any)[methodName](...args);
  }

  public async assertExchange(
    ...args: Parameters<ConfirmChannel['assertExchange']>
  ) {
    return this.seamlessCallOnChannel('assertExchange', ...args);
  }

  public async checkExchange(
    ...args: Parameters<ConfirmChannel['checkExchange']>
  ) {
    return this.seamlessCallOnChannel('checkExchange', ...args);
  }

  public async bindExchange(
    ...args: Parameters<ConfirmChannel['bindExchange']>
  ) {
    return this.seamlessCallOnChannel('bindExchange', ...args);
  }

  public async unbindExchange(
    ...args: Parameters<ConfirmChannel['unbindExchange']>
  ) {
    return this.seamlessCallOnChannel('unbindExchange', ...args);
  }

  public async deleteExchange(
    ...args: Parameters<ConfirmChannel['deleteExchange']>
  ) {
    return this.seamlessCallOnChannel('deleteExchange', ...args);
  }

  public async forceExchange(
    ...args: Parameters<ConfirmChannel['assertExchange']>
  ) {
    try {
      return await this.assertExchange(...args);
    } catch (error) {
      // https://www.rabbitmq.com/amqp-0-9-1-reference.html#constants
      if (typeof error === 'object' && [405, 406].includes(error.code)) {
        await this.deleteExchange(args[0]);

        return await this.assertExchange(...args);
      }

      throw error;
    }
  }

  public async assertQueue(...args: Parameters<ConfirmChannel['assertQueue']>) {
    return this.seamlessCallOnChannel('assertQueue', ...args);
  }

  public async checkQueue(...args: Parameters<ConfirmChannel['checkQueue']>) {
    return this.seamlessCallOnChannel('checkQueue', ...args);
  }

  public async bindQueue(...args: Parameters<ConfirmChannel['bindQueue']>) {
    return this.seamlessCallOnChannel('bindQueue', ...args);
  }

  public async unbindQueue(...args: Parameters<ConfirmChannel['unbindQueue']>) {
    return this.seamlessCallOnChannel('unbindQueue', ...args);
  }

  public async purgeQueue(...args: Parameters<ConfirmChannel['purgeQueue']>) {
    return this.seamlessCallOnChannel('purgeQueue', ...args);
  }

  public async deleteQueue(...args: Parameters<ConfirmChannel['deleteQueue']>) {
    return this.seamlessCallOnChannel('deleteQueue', ...args);
  }

  public async forceQueue(...args: Parameters<ConfirmChannel['assertQueue']>) {
    try {
      return await this.assertQueue(...args);
    } catch (error) {
      // https://www.rabbitmq.com/amqp-0-9-1-reference.html#constants
      if (typeof error === 'object' && [405, 406].includes(error.code)) {
        await this.deleteQueue(args[0]);

        return await this.assertQueue(...args);
      }

      throw error;
    }
  }
}
